[fix] iterate over reference and only copy the variable if needed

Rechi opened the pull request:
## Description
This reduces the amount of copies after f34a6f7 (#14426), which causes slow directory listings.

## Types of change
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

boba23 commented:
Just tested your commit, doesn't fix the slow dir listing for me on self compiled build.

Rechi commented:
@boba23 can you bisect which commit in #14426 causes the slow directory listing?

fritsch commented:
This is not needed at all. You build from source, just use git for bisecting. See: https://git-scm.com/docs/git-bisect

boba23 commented:
As stated, will see what I can find out tonight if no one else is able to till then ....

boba23 commented:
@fritsch @Rechi 

bear with me, since I am bisecting for the first time, but after following howtos I hope I got it right, I started bisect like this:

git bisect start 8f18c6264d93dffcc631e90d258c65178d756875 cf9c3cd3a79d0204293020de969e1d6fa002ee7d (first hash is the first comit of rechi, the other one is #14424 just before rechis comit)

first build I did was still good -> git bisect good
second build already had the slow dir listing behaviour -> git bisect bad

So I ended up with final bisect output:

2a12e8275896b249ca9cabaf80f0ba66a9fe8e4b is the first bad commit
commit 2a12e8275896b249ca9cabaf80f0ba66a9fe8e4b
Author: ace20022 <ace20022@ymail.com>
Date:   Tue Sep 11 19:52:29 2018 +0200

    [gui] Show a dvd/bluray icon for dvd/bluray folders.

:040000 040000 a00ed38bf69f6343cedbe32b5969e6259e4e39b0 a4f228f49e995f222c9f2ae1664c1a09f15089ea M      xbmc

... which means the issue isn't caused by rechis commits after all, but by @ace20022 ? @ace20022 could you look into this?

fritsch commented:
Perfect! This is another find of the same commit that @ksooo investigated with his debugger!

Thanks very much also for this detailed documentation that will help others that want to help investigate.

Rechi committed:
[fix] iterate over reference and only copy the variable if needed

This reduces the amount of copies after f34a6f7, which causes slow directory listings.
